packages/@aws-cdk/tmp-toolkit-helpers/src/api/refactoring/index.ts

```diff
@@ -0,0 +1,415 @@
+import * as crypto from 'crypto';
+import { createHash } from 'node:crypto';
+import type { TypedMapping } from '@aws-cdk/cloudformation-diff';
+import {
+  formatTypedMappings as fmtTypedMappings,
+  formatAmbiguousMappings as fmtAmbiguousMappings,
+} from '@aws-cdk/cloudformation-diff';
+import type * as cxapi from '@aws-cdk/cx-api';
+import type { StackSummary } from '@aws-sdk/client-cloudformation';
+import { deserializeStructure } from '../../util';
+import type { SdkProvider } from '../aws-auth';
+import { Mode } from '../plugin';
+import { StringWriteStream } from '../streams';
+
+interface CloudFormationTemplate {
+  Resources?: {
+    [logicalId: string]: {
+      Type: string;
+      Properties?: any;
+      Metadata?: Record<string, any>;
+    };
+  };
+}
+
+export interface BasicStack {
+  readonly environment: cxapi.Environment;
+  readonly stackName: string;
+  readonly template: CloudFormationTemplate;
+}
+
+export class AmbiguityError extends Error {
+  constructor(public readonly pairs: [ResourceLocation[], ResourceLocation[]][]) {
+    super('Ambiguous resource mappings');
+  }
+
+  public paths(): [string[], string[]][] {
+    return this.pairs.map(([a, b]) => [convert(a), convert(b)]);
+
+    function convert(locations: ResourceLocation[]): string[] {
+      return locations.map((l) => l.toPath());
+    }
+  }
+}
+
+/**
+ * This class mirrors the `ResourceLocation` interface from CloudFormation,
+ * but is richer, since it has a reference to the stack object, rather than
+ * merely the stack name.
+ */
+export class ResourceLocation {
+  constructor(readonly stack: BasicStack, readonly logicalResourceId?: string) {
+  }
+
+  public toPath(): string {
+    const stack = this.stack;
+    const resource = stack.template.Resources?.[this.logicalResourceId ?? ''];
+    const result = resource?.Metadata?.['aws:cdk:path'];
+
+    if (result != null) {
+      return result;
+    }
+
+    // If the path is not available, we can use stack name and logical ID
+    return `${stack.stackName}.${this.logicalResourceId}`;
+  }
+}
+
+/**
+ * A mapping between a source and a destination location.
+ */
+export class ResourceMapping {
+  constructor(public readonly source: ResourceLocation, public readonly destination: ResourceLocation) {
+  }
+
+  public toTypedMapping(): TypedMapping {
+    return {
+      // the type is the same in both source and destination,
+      // so we can use either one
+      type: this.getType(this.source),
+      sourcePath: this.source.toPath(),
+      destinationPath: this.destination.toPath(),
+    };
+  }
+
+  private getType(location: ResourceLocation): string {
```

Comment:
This method doesn't access `this` anywhere, so it might as well be a static helper function.